### ISSUE TITLE
Ignore declared type for failed collection literal assignments

### ISSUE THREAD

**Issue by ibraheemdev** [astral-sh/ty#3777]
Similar to https://github.com/astral-sh/ty/issues/2537, the declared type context can remain in the inferred type of a collection literal even if the assignment fails:
```py
# error: Object of type `list[int | str]` is not assignable to `list[int]`
x: list[int] = reveal_type(["a"])  # revealed: list[int | str]
```

This diagnostic can be confusing, and we should instead infer `list[str]` as the RHS, ignoring the type context.
